syntax = "proto3";

package grapher;

service Grapher {
    rpc GetLineGraph(LineGraphRequest) returns (GrapherResponse);
    rpc GetPieChart(PieChartRequest) returns (GrapherResponse);
}

message LineGraphRequest {
    // Requests a line graph for FIB totals in each address family.
    repeated Metadata metadatas = 1;
    repeated TotalTime totals_time = 2;
    string copyright = 3;
}

message PieChartRequest {
    // Requests a pie chart showing subnet distribution from the total.
    repeated Metadata metadatas = 1;
    SubnetFamily subnets = 2;
    string copyright = 3;
}

message TotalTime {
    // v4 FIB, v6 FIB, and the time it was at those values.
    uint32 v4_values = 1;
    uint32 v6_values = 2;
    uint64 time = 3;
}

message SubnetFamily {
    repeated uint32 v4_values = 1;
    repeated uint32 v6_values = 2;
}

message Metadata {
    // Used to create the image.
    string title = 1;
    uint32 x_axis = 2;
    uint32 y_axis = 3;
    uint32 colour = 4;
    repeated string colours = 5;
    repeated string labels = 6;
}

message GrapherResponse {
    // There should always be two images sent back, once for each address family.
    repeated Image images = 1;
}

message Image {
    // The image itself. Also the title of each if required.
    bytes image = 1;
    string title = 2;
}